### PR TITLE
Add support for passing through extended text attributes, like…

### LINKED ISSUE #2554: ConPTY swallows SGR 3 and 23, so terminals like Alacritty are prevented from supporting italics
Windows 10 version 1903, build 10.0.18362.295. Alacritty master.

Using Alacritty as the terminal (since it supports bold and italic text rendering), italic doesn’t work, because it seems to gets swallowed by ConPTY. https://github.com/jwilm/alacritty/issues/2745 has some discussion.

# Steps to reproduce

Run the following in Bash or similar under Alacritty, with the `enable_experimental_conpty_backend` setting enabled:

```sh
echo -e 'Normal, \x1b[1mbold\x1b[22m, \x1b[3mitalic\x1b[23m, \x1b[1;3mbold italic\x1b[22;23m'
```

(If you don’t have that setting enabled, it’ll use WinPTY which also doesn’t pass italics through, but that’s irrelevant here.)

# Expected behavior

Output like this:

> Normal, **bold**, *italic*, ***bold italic***

# Actual behavior

Italics don’t occur:

> Normal, **bold**, italic, **bold italic**

Further, when I insert logging into Alacritty’s `csi_dispatch` method, to observe what SGR codes it’s actually hearing about and applying, I see that the 3 (italic) and 23 (stop italic) have been swallowed, presumably by ConPTY, and Alacritty never saw them.

(ConPTY also modifies the `CSI [ 1 m` bold code to add another SGR code to change to the high-intensity colour. I still haven’t decided how I feel about that!)

I don’t know much about how these things work, so I don’t know if this is normal, but I was caught by surprise that ConPTY was apparently rendering things before they ever got to Alacritty, so that terminals are limited in what they can implement by what ConPTY implements. I expected it to pass everything through untouched, or at least to pass unknown things through rather than swallowing them.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request
Adds support for Italics, Blinking, Invisible, CrossedOut text, THROUGH CONPTY. This does **NOT** add support for those styles to conhost or the terminal.

We will store these "Extended Text Attributes" in a `TextAttribute`. When we go to render a line, we'll see if the state has changed from our previous state, and if so, we'll appropriately toggle that state with VT. Boldness has been moved from a `bool` to a single bit in these flags.

Technically, now that these are stored in the buffer, we only need to make changes to the renderers to be able to support them. That's not being done as a part of this PR however.

## References
See also #2915 and #2916, which are some follow-up tasks from this fix. I thought them too risky for 20H1.

## PR Checklist
* [x] Closes #2554
* [x] I work here
* [x] Tests added/passed
* [n/a] Requires documentation to be updated

## Validation Steps Performed
* Replaced the conhost in my system32 and tried running the following script, via WSL interop:

<details>

```python
import sys
import time # time.sleep is in seconds
def write(s):
    sys.stdout.write(s)

def csi(seq):
    sys.stdout.write('\x1b[{}'.format(seq))

def sgr(code=0):
    csi('{}m'.format(code))

def cupxy(x=0, y=0):
    cup(y+1, x+1)

def clear_all():
    cupxy(0,0)
    csi('2J')

# $ python ~/vttests/extended-attrs.py
if __name__ == '__main__':
    clear_all()
    print('This is the VT Test template.')

    sgr(0)
    write('[Normal Text]')
    sgr(3)
    write('[Italicized]')
    sgr(0)
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(4)
    write('[Underlined]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(4)
    write('[Underlined]')
    sgr(3)
    write('[U & I]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(4)
    write('[Underlined]')
    sgr(3)
    write('[U & I]')
    sgr(23)
    write('[Underlined]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(4)
    write('[Underlined]')
    sgr(3)
    write('[U & I]')
    sgr(24)
    write('[Italicized]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(5)
    write('[Blinking]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(8)
    write('[Invisible]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(9)
    write('[Crossed-out]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    ############################################################################


    sgr(0)
    write('[Normal Text]')
    sgr(1)
    write('[Bold]')
    sgr(3)
    write('[B & Italicized]')
    sgr(0)
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(1)
    write('[Bold]')
    sgr(4)
    write('[B & Underlined]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(1)
    write('[Bold]')
    sgr(4)
    write('[B & Underlined]')
    sgr(3)
    write('[B & U & I]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(1)
    write('[Bold]')
    sgr(4)
    write('[B & Underlined]')
    sgr(3)
    write('[U & I]')
    sgr(23)
    write('[B & Underlined]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(1)
    write('[Bold]')
    sgr(4)
    write('[B & Underlined]')
    sgr(3)
    write('[B & U & I]')
    sgr(24)
    write('[B & Italicized]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(1)
    write('[Bold]')
    sgr(5)
    write('[B & Blinking]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(1)
    write('[Bold]')
    sgr(8)
    write('[B & Invisible]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

    sgr(0)
    write('[Normal Text]')
    sgr(1)
    write('[Bold]')
    sgr(9)
    write('[B & Crossed-out]')
    sgr(0)
    write('[Normal Again]')
    write('\n')

```

</details>

* ran tests
* Using the (added) static assert, ensured that the size of the TextAttribute did not increase.

### CODE REVIEW

File: src/renderer/vt/VtSequences.cpp [hunk 1]
```diff
@@ -345,3 +345,91 @@ using namespace Microsoft::Console::Render;
...
+
+// Method Description:
+// - Writes a sequence to tell the terminal to start italicizing text
+// Arguments:
+// - <none>
+// Return Value:
+// - S_OK if we succeeded, else an appropriate HRESULT for failing to allocate or write.
+[[nodiscard]] HRESULT VtEngine::_BeginItalics() noexcept
+{
+    return _Write("\x1b[3m");
```
Comment (DHowett-MSFT) [inline src/renderer/vt/VtSequences.cpp]: I wonder if there's an optimization we can take later that'll let us compress SGRs together when we know we're writing a bunch. `\x1b[40m`+`\x1b[31m`+`\x1b[4m`+`\x1b[3m` => `\x1b[40;31;4;3m`?

File: src/renderer/vt/Xterm256Engine.cpp [hunk 2]
```diff
@@ -42,11 +43,77 @@ Xterm256Engine::Xterm256Engine(_In_ wil::unique_hfile hPipe,
...
+[[nodiscard]] HRESULT Xterm256Engine::_UpdateExtendedAttrs(const ExtendedAttributes extendedAttrs) noexcept
+{
+    // Helper lambda to check if a state (attr) has changed since it's last
+    // value (lastState), and appropriately start/end that state with the given
+    // begin/end functions.
+    auto updateFlagAndState = [extendedAttrs](const ExtendedAttributes attr,
+                                              bool& lastState,
+                                              std::function<HRESULT(void)> beginFn,
+                                              std::function<HRESULT(void)> endFn) -> HRESULT {
+        // Unfortunately, we can't use WI_IsFlagSet here, since that requires a
```
Comment (DHowett-MSFT) [inline src/renderer/vt/Xterm256Engine.cpp]: I thought there was another helper for this

File: src/renderer/vt/Xterm256Engine.cpp [hunk 3]
```diff
@@ -42,11 +43,77 @@ Xterm256Engine::Xterm256Engine(_In_ wil::unique_hfile hPipe,
...
+                RETURN_IF_FAILED(endFn());
+            }
+            lastState = flagSet;
+        }
+        return S_OK;
+    };
+
+    auto hr = updateFlagAndState(ExtendedAttributes::Italics,
+                                 _usingItalics,
+                                 std::bind(&Xterm256Engine::_BeginItalics, this),
```
Comment (DHowett-MSFT) [inline src/renderer/vt/Xterm256Engine.cpp]: binds are sooooooooo expensive. :|

Comment (DHowett-MSFT) [reply src/renderer/vt/Xterm256Engine.cpp]: i wonder if there's a trick you can do with auto-typed args in lambdas

File: src/terminal/adapter/DispatchTypes.hpp [hunk 4]
```diff
@@ -13,21 +13,30 @@ namespace Microsoft::Console::VirtualTerminal::DispatchTypes
         Scrollback = 3
     };
 
+    // TODO:GH#2916 add support for DoublyUnderlined, Faint(2) to the adapter as well.
     enum GraphicsOptions : unsigned int
     {
         Off = 0,
         BoldBright = 1,
-        RGBColor = 2,
+        RGBColorOrFaint = 2,
```
Comment (DHowett-MSFT) [inline src/terminal/adapter/DispatchTypes.hpp]: Oh, I see: `38;2` versus `2`. stupid vt encoding!

File: src/terminal/adapter/DispatchTypes.hpp [hunk 5]
```diff
@@ -13,21 +13,30 @@ namespace Microsoft::Console::VirtualTerminal::DispatchTypes
...
-        RGBColor = 2,
+        RGBColorOrFaint = 2,
         // 2 is also Faint, decreased intensity (ISO 6429).
+        Italics = 3,
         Underline = 4,
-        Xterm256Index = 5,
+        BlinkOrXterm256Index = 5,
         // 5 is also Blink (appears as Bold).
         // the 2 and 5 entries here are only for the extended graphics options
         // as we do not currently support those features individually
```
Comment (DHowett-MSFT) [inline src/terminal/adapter/DispatchTypes.hpp]: comment technically out of date

File: src/terminal/adapter/DispatchTypes.hpp [hunk 6]
```diff
@@ -13,21 +13,30 @@ namespace Microsoft::Console::VirtualTerminal::DispatchTypes
...
+        CrossedOut = 9,
+        DoublyUnderlined = 21,
         UnBold = 22,
+        NotItalics = 23,
         NoUnderline = 24,
-        Positive = 27,
+        Steady = 25, // _not_ blink
+        Positive = 27, // _not_ inverse
+        Visible = 28, // _not_ invisible
+        NotCrossedOut = 29, // _not_ invisible
```
Comment (DHowett-MSFT) [inline src/terminal/adapter/DispatchTypes.hpp]:
```suggestion
        NotCrossedOut = 29, // _not_ crossed out
```

Comment (DHowett-MSFT) [reply src/terminal/adapter/DispatchTypes.hpp]: does the comment add value here?

File: src/terminal/adapter/adaptDispatchGraphics.cpp [hunk 7]
```diff
@@ -424,14 +528,21 @@ bool AdaptDispatch::SetGraphicsRendition(_In_reads_(cOptions) const DispatchType
...
+                                               &fIsForeground,
+                                               &cOptionsConsumed);
 
                 i += (cOptionsConsumed - 1); // cOptionsConsumed includes the opt we're currently on.
             }
             else
             {
                 _SetGraphicsOptionHelper(opt, &attr);
-                fSuccess = !!_conApi->PrivateSetLegacyAttributes(attr, _fChangedForeground, _fChangedBackground, _fChangedMetaAttrs);
+                fSuccess = !!_conApi->PrivateSetLegacyAttributes(attr,
```
Comment (DHowett-MSFT) [inline src/terminal/adapter/adaptDispatchGraphics.cpp]: this won't stomp ext attrs, will it?

File: src/host/ut_host/ScreenBufferTests.cpp [hunk 8]
```diff
@@ -4403,3 +4405,131 @@ void ScreenBufferTests::InitializeTabStopsInVTMode()
...
+    }
+    if (crossedOut)
+    {
+        WI_ClearFlag(expectedAttrs, ExtendedAttributes::CrossedOut);
+        vtSeq = L"\x1b[29m";
+        validate(expectedAttrs, vtSeq);
+    }
+
+    stateMachine.ProcessString(L"\x1b[0m");
+}
```
Comment (DHowett-MSFT) [inline src/host/ut_host/ScreenBufferTests.cpp]: i'd be interested in a test that validates that separate sequences that set colors and then extended attributes results in the correct rendition

File: src/host/getset.cpp [hunk 9]
```diff
@@ -970,6 +970,37 @@ void DoSrvPrivateBoldText(SCREEN_INFORMATION& screenInfo, const bool bolded)
     buffer.SetAttributes(attrs);
 }
 
+// Method Description:
+// - Retrieves the active ExtendedAttributes of the given screen buffer. Text
```
Comment (miniksa) [inline src/host/getset.cpp]: Maybe leave an example of what a few "extended attributes" are here and below to save someone the effort of looking it up. I often say "Attributes, like colors" when writing such comments.

File: src/host/outputStream.hpp [hunk 10]
```diff
@@ -88,6 +88,8 @@ class ConhostInternalGetSet final : public Microsoft::Console::VirtualTerminal::
                                     const bool fIsForeground) override;
 
     BOOL PrivateBoldText(const bool bolded) override;
+    BOOL PrivateGetExtendedTextAttributes(ExtendedAttributes* const pAttrs) override;
```
Comment (miniksa) [inline src/host/outputStream.hpp]: Why not just a ref? Matching existing pattern?

File: src/host/ut_host/ScreenBufferTests.cpp [hunk 11]
```diff
@@ -4403,3 +4405,131 @@ void ScreenBufferTests::InitializeTabStopsInVTMode()
 
     VERIFY_IS_TRUE(gci.GetActiveOutputBuffer().AreTabsSet());
 }
+
+void ScreenBufferTests::TestExtendedTextAttributes()
+{
+    // This is a test for microsoft/terminal#2554. Refer to that issue for more
```
Comment (miniksa) [inline src/host/ut_host/ScreenBufferTests.cpp]: I mean, give me a little summary of it here.

File: src/renderer/vt/WinTelnetEngine.cpp [hunk 12]
```diff
@@ -29,17 +29,22 @@ WinTelnetEngine::WinTelnetEngine(_In_ wil::unique_hfile hPipe,
 // - colorBackground: The RGB Color to use to paint the background of the text.
 // - legacyColorAttribute: A console attributes bit field specifying the brush
 //      colors we should use.
+// - extendedAttrs - extended text attributes (italic, underline, etc.) to use.
```
Comment (miniksa) [inline src/renderer/vt/WinTelnetEngine.cpp]: Yeah, like this one! Do this where you mention "extended attrs"!

File: src/renderer/vt/Xterm256Engine.cpp (hunk 2 above)
Comment (miniksa) [reply src/renderer/vt/Xterm256Engine.cpp]: `WI_AreAllFlagsSet`

File: src/renderer/vt/Xterm256Engine.cpp (hunk 3 above)
Comment (miniksa) [reply src/renderer/vt/Xterm256Engine.cpp]: If we have a particular qualm with `std::bind`, we should probably file a code health work item with the proposed replacement pattern for them. There's a ton of existing ones.

File: src/terminal/adapter/adaptDispatchGraphics.cpp [hunk 13]
```diff
@@ -12,6 +12,19 @@
 using namespace Microsoft::Console::VirtualTerminal;
 using namespace Microsoft::Console::VirtualTerminal::DispatchTypes;
 
+// Inspired from RETURN_IF_WIN32_BOOL_FALSE
```
Comment (miniksa) [inline src/terminal/adapter/adaptDispatchGraphics.cpp]: This should go into a more public helper section than just the header of this file and also link to an upstream contribution to Microsoft/wil if you've indeed identified a gap here.

File: src/terminal/adapter/adaptDispatchGraphics.cpp [hunk 14]
```diff
@@ -261,6 +279,27 @@ void AdaptDispatch::_SetGraphicsOptionHelper(const DispatchTypes::GraphicsOption
...
+// Routine Description:
+// Returns true if the GraphicsOption represents an extended text attribute.
+//   These include things such as Underlined, Italics, Blinking, etc.
+// Return Value:
+// - true if the opt is the indicator for an extended text attribute, false otherwise.
+bool AdaptDispatch::s_IsExtendedTextAttribute(const DispatchTypes::GraphicsOptions opt) noexcept
+{
+    // TODO:GH#2916 add support for DoublyUnderlined, Faint(RGBColorOrFaint).
+    // These two are currently partially implemented as other things:
+    // * Faint is kinda like the opposite of what bold does
```
Comment (miniksa) [inline src/terminal/adapter/adaptDispatchGraphics.cpp]:
```suggestion
    // * Faint is approximately what the opposite of what bold does
```

"kinda like"... ehhhhhhhhh

File: src/terminal/adapter/DispatchTypes.hpp (hunk 6 above)
Comment (zadjii-msft) [reply src/terminal/adapter/DispatchTypes.hpp]: I guess that one is pretty straight forward

File: src/host/outputStream.hpp (hunk 10 above)
Comment (zadjii-msft) [reply src/host/outputStream.hpp]: That's correct, I was just trying to match the existing interface here. It's a little dated sure, but it's _consistent_

File: src/terminal/adapter/adaptDispatchGraphics.cpp (hunk 13 above)
Comment (zadjii-msft) [reply src/terminal/adapter/adaptDispatchGraphics.cpp]: I've filed microsoft/wil#98 upstream to discuss. I'll move this somewhere more general

File: src/renderer/vt/Xterm256Engine.cpp (hunk 3 above)
Comment (zadjii-msft) [reply src/renderer/vt/Xterm256Engine.cpp]: I just did a cursory find-all for `std::bind`, and it didn't look like there were any other perf-intensive (loops) places where we did a bind, usually only on startup. @DHowett-MSFT think we should still file a follow-up? I was able to remove this one at least.

File: src/renderer/vt/VtSequences.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/renderer/vt/VtSequences.cpp]: (moved to #3016)

**Comment by DHowett-MSFT** [comment]
You may need to merge in master to get the CI pipeline repaired.

File: src/buffer/out/TextAttribute.hpp [hunk 15]
```diff
@@ -53,7 +55,8 @@ class TextAttribute final
         _wAttrLegacy{ 0 },
         _foreground{ rgbForeground },
         _background{ rgbBackground },
-        _isBold{ false }
+        // _isBold{ false }
```
Comment (carlos-zamora) [inline src/buffer/out/TextAttribute.hpp]:
```suggestion
```

File: tools/bcz.cmd [hunk 16]
```diff
@@ -68,7 +68,7 @@ if "%_EXCLUSIVE%" == "1" (
 echo Performing nuget restore...
 nuget.exe restore %OPENCON%\OpenConsole.sln
 
-set _BUILD_CMDLINE="%MSBUILD%" %OPENCON%\OpenConsole.sln /t:%_MSBUILD_TARGET% /m /p:Configuration=%_LAST_BUILD_CONF% /p:Platform=%ARCH% %_APPX_ARGS%
+set _BUILD_CMDLINE="%MSBUILD%" %OPENCON%\OpenConsole.sln /t:"%_MSBUILD_TARGET%" /m /p:Configuration=%_LAST_BUILD_CONF% /p:Platform=%ARCH% %_APPX_ARGS%
```
Comment (carlos-zamora) [inline tools/bcz.cmd]: wat? why?

File: src/renderer/vt/Xterm256Engine.hpp [hunk 17]
```diff
@@ -35,10 +35,17 @@ namespace Microsoft::Console::Render
...
                                                    const COLORREF colorBackground,
                                                    const WORD legacyColorAttribute,
-                                                   const bool isBold,
+                                                   const ExtendedAttributes extendedAttrs,
                                                    const bool isSettingDefaultBrushes) noexcept override;
 
     private:
+        [[nodiscard]] HRESULT _UpdateExtendedAttrs(const ExtendedAttributes extendedAttrs) noexcept;
+
+        bool _usingItalics{ false };
```
Comment (carlos-zamora) [inline src/renderer/vt/Xterm256Engine.hpp]: Move to initializer list?

File: src/renderer/vt/Xterm256Engine.cpp [hunk 18]
```diff
@@ -42,11 +43,77 @@ Xterm256Engine::Xterm256Engine(_In_ wil::unique_hfile hPipe,
...
+[[nodiscard]] HRESULT Xterm256Engine::_UpdateExtendedAttrs(const ExtendedAttributes extendedAttrs) noexcept
+{
+    // Helper lambda to check if a state (attr) has changed since it's last
+    // value (lastState), and appropriately start/end that state with the given
+    // begin/end functions.
+    auto updateFlagAndState = [extendedAttrs, this](const ExtendedAttributes attr,
+                                                    bool& lastState,
+                                                    std::function<HRESULT(Xterm256Engine*)> beginFn,
+                                                    std::function<HRESULT(Xterm256Engine*)> endFn) -> HRESULT {
+        // Unfortunately, we can't use WI_IsFlagSet here, since that requires a
```
Comment (carlos-zamora) [inline src/renderer/vt/Xterm256Engine.cpp]: Remove old comment

File: tools/bcz.cmd (hunk 16 above)
Comment (zadjii-msft) [reply tools/bcz.cmd]: The _MSBUILD_TARGET can have spaces in it, and if it does, this won't work.Case in point: any project in the "Virtual Terminal" solution directory. Wrapping it with quotes fixes that.

File: src/buffer/out/TextAttribute.hpp [hunk 19]
```diff
@@ -169,6 +182,13 @@ class TextAttribute final
 #endif
 };
 
+#pragma pack(pop)
+// 2 for _wAttrLegacy
+// 4 for _foreground
+// 4 for _foreground
```
Comment (miniksa) [inline src/buffer/out/TextAttribute.hpp]:
```suggestion
// 4 for _background
```

File: src/renderer/vt/Xterm256Engine.cpp [hunk 20]
```diff
@@ -42,11 +47,74 @@ Xterm256Engine::Xterm256Engine(_In_ wil::unique_hfile hPipe,
...
+        return S_OK;
+    };
+
+    auto hr = updateFlagAndState(ExtendedAttributes::Italics,
+                                 _usingItalics,
+                                 &Xterm256Engine::_BeginItalics,
+                                 &Xterm256Engine::_EndItalics);
+    RETURN_IF_FAILED(hr);
+
+    hr = updateFlagAndState(ExtendedAttributes::Blinking,
```
Comment (miniksa) [inline src/renderer/vt/Xterm256Engine.cpp]: Nnnn... what happens in the scenario where:

- We're attempting to set Italics, Blink, and Invisible.
- Italics succeeds
- Blink fails
- Invisible would have succeeded, but we early returned...

Are we OK with that happening or should we have a provision to try the others and aggregate the result code later?

I don't think there's really a way for us to make this atomic where they all work or none of them work... or is there?

EDIT: I'm worried that the potentially torn state will result in a weird debugging situation in the future...

File: src/renderer/vt/Xterm256Engine.hpp [hunk 21]
```diff
@@ -35,10 +35,17 @@ namespace Microsoft::Console::Render
...
                                                    const COLORREF colorBackground,
                                                    const WORD legacyColorAttribute,
-                                                   const bool isBold,
+                                                   const ExtendedAttributes extendedAttrs,
                                                    const bool isSettingDefaultBrushes) noexcept override;
 
     private:
+        [[nodiscard]] HRESULT _UpdateExtendedAttrs(const ExtendedAttributes extendedAttrs) noexcept;
+
+        bool _usingItalics;
```
Comment (miniksa) [inline src/renderer/vt/Xterm256Engine.hpp]: This feels like flags to me (or enum flags). Why four bools instead?

File: src/terminal/adapter/adaptDispatchGraphics.cpp [hunk 22]
```diff
@@ -12,6 +12,19 @@
...
+// Inspired from RETURN_IF_WIN32_BOOL_FALSE
+// WIL doesn't include a RETURN_IF_FALSE, and RETURN_IF_WIN32_BOOL_FALSE
+//  will actually return the value of GLE.
+#define RETURN_IF_FALSE(b)                    \
+    do                                        \
+    {                                         \
+        BOOL __boolRet = wil::verify_bool(b); \
+        if (!__boolRet)                       \
+        {                                     \
+            return b;                         \
```
Comment (miniksa) [inline src/terminal/adapter/adaptDispatchGraphics.cpp]:
```suggestion
            return __boolRet;                         \
```

Per what Duncan Horn said in the upstream issue.

File: src/renderer/vt/Xterm256Engine.cpp (hunk 20 above)
Comment (zadjii-msft) [reply src/renderer/vt/Xterm256Engine.cpp]: You're right that's it's possible that we'd end in a bad state if one of these fails. Looking at the implementation of these, this is the only spot that can actually fail:
```c++
    try
    {
        _buffer.append(str);
        return S_OK;
    }
    CATCH_RETURN();
```
(where `str` is the VT sequence we're writing). IMO, if that fails, then we're probably going to end up in a much worse torn state than just this one sequence being lost.

The idea to do them all atomically would probably match well with #3016.

Comment (miniksa) [reply src/renderer/vt/Xterm256Engine.cpp]: Let's stash the idea inside the notes for #3016 then and just let this roll as is.

**Pull request merged**